dateutil/test/test_parser.py

```diff
@@ -238,6 +238,60 @@ def test_parse_bytearray(self):
         assert res == expected
 
 
+class TestTzinfoInputTypes(object):
+    def test_tzinfo_dict_could_return_none(self):
+        dstr = "2017-02-03 12:40 BRST"
+        result = parse(dstr, tzinfos={"BRST": None})
+        expected = datetime(2017, 2, 3, 12, 40)
+        assert result == expected
+
+    def test_tzinfos_callable_could_return_none(self):
+        dstr = "2017-02-03 12:40 BRST"
+        result = parse(dstr, tzinfos=lambda *args: None)
+        expected = datetime(2017, 2, 3, 12, 40)
+        assert result == expected
+
+    def test_invalid_tzinfo_input(self):
+        dstr = "2014 January 19 09:00 UTC"
+        # Pass an absurd tzinfos object
+        tzinfos = {"UTC": ValueError}
+        with pytest.raises(TypeError):
+            parse(dstr, tzinfos=tzinfos)
+
+    def test_valid_tzinfo_tzinfo_input(self):
+        dstr = "2014 January 19 09:00 UTC"
+        tzinfos = {"UTC": tz.tzutc()}
+        expected = datetime(2014, 1, 19, 9, tzinfo=tz.tzutc())
+        res = parse(dstr, tzinfos=tzinfos)
+        assert res == expected
+
+    def test_valid_tzinfo_unicode_input(self):
+        dstr = "2014 January 19 09:00 UTC"
+        tzinfos = {u"UTC": u"UTC+0"}
+        expected = datetime(2014, 1, 19, 9, tzinfo=tz.tzutc())
+        res = parse(dstr, tzinfos=tzinfos)
+        assert res == expected
+
+    def test_valid_tzinfo_callable_input(self):
+        dstr = "2014 January 19 09:00 UTC"
+
+        def tzinfos(*args, **kwargs):
+            return u"UTC+0"
+
+        expected = datetime(2014, 1, 19, 9, tzinfo=tz.tzutc())
+        res = parse(dstr, tzinfos=tzinfos)
+        assert res == expected
+
+    @pytest.mark.xfail(reason="Doesn't check for UTC in tzinfos dict, "
```

Comment:
This is a weird test but I think you're right that it should be possible.

I think a more plausible use of this functionality would be something like returning a custom UTC type, e.g. `datetime.timezone.utc`, but for now let's leave it as is.